### PR TITLE
Internals re-implementation

### PR DESCRIPTION
**Pull request by lorenzo**
Adding new features to Phinx was proving difficult for myself, due to its current arquitechture.

I decided to re-implement most of the internals as an attempt to remedy the situation and open up a way to add several more improvements in the future. I'll start with the benefits I was able to extract for this new architecture:

## Benefits so far

* Migrations run significantly faster: Instructions are bundled into as little `ALTER` instructions as it can do
* Separation of fast VS slow instructions: Things like adding indexes and creating foreign keys are differed to be run last. This opens the way for executing the instructions concurrently in the database for even faster migrations.

## The new architecture

Running migrations is very similar to compiling and executing a program. The executing a php file, for example, the parser will transform the instruction into an intermediate structure and then it executes it as it reads it.Optionally, php can also re-order some parts of the intermediate structure to add some optimisations.

This new architecture fits the same idea:

* The migration files are the programming language
* Phinx is the compiler
* The adapters are the machines running the program

The problem with the old architecture was that instructions were sent immediately to the machine, instead of creating the intermediate structure. This new architecture builds an `Intent`, which is then transformed and optimised in several steps before it is finally sent to the database.

The new program flow is in broad terms as follows:

* Migration files internally create an `Intent` object containing all `Action` objects to execute
* `Action` objects are just value objects representing something to do in the database
* A `Plan` object is used to transform the `Intent` into a set of intermediate structures: Either `NewTable` or `AlterTable` objects
* The plan steps are then delegated to the adapters, which once again transform the `AlterTable` objects into another object called `AlterInstructions`
* `AlterInstructions` are then compiled down to SQL and executed one by one in the database.

As you may guess, each transformation opens the possibility of adding certain optimisations. During this refactoring I only added one of such optimisations: The ability to run all `ALTER` instructions for a table as a single command. This is specially beneficial for `Mysql` and `Postgresql`

## Breaking changes

This is a very significant breaking change, and I guess we can have that luxury as phinx has not been marked as stable. From the user point of view, there should be no visible changes, but for anyone extending the internals, this will break their code for sure.

### CODE REVIEW

File: src/Phinx/Db/Action/Action.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+namespace Phinx\Db\Action;
+
+abstract class Action
```
Comment (rquadling) [inline src/Phinx/Db/Action/Action.php]: Is there a convention of having an 'Abstract' prefix to abstract classes? Also,by not having a constructor, this is placing the responsibility of populating the $table on the subclass with no type hinting being enforced in this class. Making $table private and having a `protected function setTable(Table $table){$this->table=$table;}` would provide that enforcement of the correct type. Or having a constructor and having various tools (PHPStan for example) to identify when/if the subclass doesn't call the parent class

File: src/Phinx/Db/Action/AddColumn.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ */
+namespace Phinx\Db\Action;
+
+use Phinx\Db\Table\Column;
+use Phinx\Db\Table\Table;
+
+class AddColumn extends Action
+{
+
+    protected $column;
```
Comment (rquadling) [inline src/Phinx/Db/Action/AddColumn.php]: Typehint

File: src/Phinx/Db/Action/AddColumn.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    public function __construct(Table $table, Column $column)
+    {
+        $this->table = $table;
+        $this->column = $column;
+    }
+
+    public static function build(Table $table, $columnName, $type = null, $options = [])
```
[the remark this answers is not in the document]
Comment (lorenzo) [reply src/Phinx/Db/Action/AddColumn.php]: Yeah, I don’t see why not. I see it similar to overloading the constructor in java, only that php has no method overloading.

File: src/Phinx/Db/Action/Action.php (hunk 1 above)
Comment (lorenzo) [reply src/Phinx/Db/Action/Action.php]: I’ll add a constructor, I don’t want setters in the action classes

File: src/Phinx/Db/Action/AddColumn.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (rquadling) [reply src/Phinx/Db/Action/AddColumn.php]: I'd have an ActionInterface that requires each Action to be able to build itself then.

[the remark this answers is not in the document]
Comment (rquadling) [reply src/Phinx/Db/Action/AddColumn.php]: Or an ActionFactory that builds one based upon the supplied data.

[the remark this answers is not in the document]
Comment (lorenzo) [reply src/Phinx/Db/Action/AddColumn.php]: Do you have any reason in mind for that?

File: src/Phinx/Db/Adapter/DirectActionInterface.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     * @return void
+     */
+    public function addForeignKey(Table $table, ForeignKey $foreignKey);
+
+    /**
+     * Drops the specified foreign key from a database table.
+     *
+     * @param string   $tableName
+     * @param string[] $columns    Column(s)
+     * @param string   $constraint Constraint name
```
Comment (dereuromark) [inline src/Phinx/Db/Adapter/DirectActionInterface.php]: |null here

File: src/Phinx/Db/Action/AddColumn.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (rquadling) [reply src/Phinx/Db/Action/AddColumn.php]: For an object whose main intention is to be passed around once created, itself acting as its own factory is redundant. Once instantiated the objects only purpose is to act as a value object. Not the one hit wonder factory.

The single responsibility aspect is what comes to mind here. Either the object represents some state OR it creates something stateful.

I may be being picky, but if the refactoring is taken in, then a very clear separation would be nicer. Smaller objects, doing fewer things. Far easier to test and understand. But as in all things, compromise between a purists perspective and a practical one.

**Comment by ravage84** [comment]
> This is a very significant breaking change, and I guess we can have that luxury as phinx has not been marked as stable. From the user point of view, there should be no visible changes, but for anyone extending the internals, this will break their code for sure.

Agree,. Let's do anything that seems necessary to bring Phinx to a stable & robust base for the intermediate future.
In the long run we can always introduce more breaking changes that demand a major bump once we have enough experience, though I'd recommend us to have a plan for easing the upgrade/migration (remember CakePHP 2.x -> 3.x).

**Pull request merged**